### PR TITLE
First attempt

### PR DESCRIPTION
**Pull request by howardthegr8one**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -13,3 +13,23 @@
...
+
+Question 2)
+
+- Since the runtime to traverse a binary search tree is $log_2(n)$ we can use this to estimate the difference between $log_2(1000)$ and $log_2(10000)$.
+
+- $log_2(1000)$ is approximately 9.966 and $log_2(10000)$ is approximately 13.288, thus 13.288/9.966 is roughly 1.333 repeating. Using these values we can conclude that if the time to search 1000 elements is 5 seconds, then the time to search 10000 would be 5 * 1.333 or 6.666 seconds as that is the growth rate we found from $log_2(1000)$ to $log_2(10000)$
+
+Question 3)
+
+- The algorithm might be implemented in such a way that the slowness isn't apparent until the input size is large enough. For example a brute force implementation to solve the traveling salesperson problem will run fine with an input smaller than say 10, but attempting to search among 100 or even 50 and the algorithm will likely run out of memory before finishing. Perhaps this binary search algorithm is implemented similarly.
```
Comment (larskotthoff) [inline README.md]: But the asymptotic complexity tells us about exactly this case -- what happens when the input is large.

File: README.md [hunk 2]
```diff
@@ -13,3 +13,23 @@
...
+
+Question 3)
+
+- The algorithm might be implemented in such a way that the slowness isn't apparent until the input size is large enough. For example a brute force implementation to solve the traveling salesperson problem will run fine with an input smaller than say 10, but attempting to search among 100 or even 50 and the algorithm will likely run out of memory before finishing. Perhaps this binary search algorithm is implemented similarly. 
+- It could simply be a hardware issue, or different data sets as input. If the programmer tested the runtime for 1000 elements with a different input than when testing 10000 elements this could account for such a large change in runtime, or merely testing in one environment, and then testing again in a different environment.
+- The data structure the algorithm uses doesn't scale well with larger inputs. Even if the algorithm logically works, an ineffecient choice of data structure could be increasing the runtime beyond what the typical traversal of a binary tree suggests.
```
Comment (larskotthoff) [inline README.md]: But then it wouldn't have that asymptotic complexity.

**Review by larskotthoff — CHANGES_REQUESTED** [review]

File: README.md [hunk 3]
```diff
@@ -30,6 +30,8 @@ Question 2)
...
-- It could simply be a hardware issue, or different data sets as input. If the programmer tested the runtime for 1000 elements with a different input than when testing 10000 elements this could account for such a large change in runtime, or merely testing in one environment, and then testing again in a different environment.
-- The data structure the algorithm uses doesn't scale well with larger inputs. Even if the algorithm logically works, an ineffecient choice of data structure could be increasing the runtime beyond what the typical traversal of a binary tree suggests.
\ No newline at end of file
+- Our time estimate was based on our assumption that the runtime complexity is $log(n)$, however the actual time it takes to run could be slower than estimated due to a constant of some kind, perhaps the actual runtime is closer to $10 log(n)$ or $100 log(n)$ which would account for a much slower runtime in practice.
```
Comment (larskotthoff) [inline README.md]: How would a constant factor change your estimate?

**Review by larskotthoff — CHANGES_REQUESTED** [review]

File: README.md [hunk 4]
```diff
@@ -13,3 +13,25 @@
...
+
+# My answers
+
+Question 1)
+
+- You could have two algorithms that both have an asymptotic runtime of $O(n^3)$ but one algorithm might have a constant factor such as $5n^3$ whereas the other algorithm might just be $n^3$. Although asymptotically both of these algorithms have the same $O(n^3)$ the one with a constant factor in practice will run much slower than the second algorithm without any constants.
+- Additionally two algorithms could both have a runtime complexity of $O(n^3)$ but exploring further you might find that one of the algorithms is $n^3$ while the other is $5n^3 + 3n^2 + 2n$ in this case the big O is the same but in practice the algorithm with extra lower order terms will run much slower than the algorithm which only runs in $O(n^3)$
+- Asymptotic analysis doesn't account for differences or limitations in hardware. Therefore you could have an algorithm with a seemingly fast asymptotic runtime that in practice can't make full use of its implementation because of hardware limitations, this could especially be true in the case of asynchronous algorithms that may not have as many threads or cores as it needs to maximize it's effeciency.
```
Comment (larskotthoff) [inline README.md:23]: ...but the asymptotic analysis we've been doing doesn't take parallelism into account at all.

File: README.md [hunk 5]
```diff
@@ -13,3 +13,25 @@
...
+
+Question 2)
+
+- Since the runtime to traverse a binary search tree is $log_2(n)$ we can use this to estimate the difference between $log_2(1000)$ and $log_2(10000)$.
+
+- $log_2(1000)$ is approximately 9.966 and $log_2(10000)$ is approximately 13.288, thus 13.288/9.966 is roughly 1.333 repeating. Using these values we can conclude that if the time to search 1000 elements is 5 seconds, then the time to search 10000 would be 5 * 1.333 or 6.666 seconds as that is the growth rate we found from $log_2(1000)$ to $log_2(10000)$
+
+Question 3)
+
+- Similarly to the first question, our estimate was based on the asymptotic runtime complexity. Therefore since the algorithm is performing slower in practice there could be lower order terms or constants that were unaccounted for in the asymptotic complexity that would account for a slower runtime in practice.
```
Comment (larskotthoff) [inline README.md:33]: But those would affect the run for the smaller input as well.

File: README.md [hunk 6]
```diff
@@ -13,3 +13,25 @@
...
+
+- When testing the algorithm with 1000 elements better hardware could've been used, or no software was running in the background, whereas when testing with 10,000 elements slower hardware with other software running could slow down the actual time beyond what we estimated from the assumed time complexity.
+
+- Lastly the type of input could also account for such a large difference in our estimate compared to the actual runtime. For example when testing with $1,000$ elements the input could've been a best-case or close to best-case input. If the target we're searching for is found after only a few recursive calls then this runtime would look faster than is expected, furthermore if our test with $10,000$ elements was a worst-case or close to worst-case scenario it would still take $log(n)$ time but it would make our estimate much smaller than what a non-best-case test input would lead us to believe.
```
Comment (larskotthoff) [inline README.md:37]: Don't really see that -- how many more levels would a binary tree have doing from 1000 to 10000?

**Review by larskotthoff — CHANGES_REQUESTED** [review]

**Comment by howardthegr8one** [comment]
Not sure why but when I went back to the correct feedback branch it created a conflict between the branches, so I'm closing this one and just going back to the github classroom feedback branch for my work.

**Pull request closed without merging**